source/custom/api/types.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+// Rule defines a rule for node customization such as labeling.
+type Rule struct {
```

Comment:
i don't like having duplicated cope like this. but for now it's ok.
I think for v0.16 I want to work on moving Rule type outside the v1alpha1 folder/pkg to clean the api pkg and also to have all that functionality in a more centralized location

Reply:
Regarding duplication, I cannot think of a better alternative. The two APIs (the CRD API and the custom source configuration) have diverged. In terms of typedefs the custom source config is a subset of the CRD API, but in terms of "syntax", it's a superset as it supports these int-or-string-or-bool kind of fields and some historical shortforms etc

We cannot move the typedef outside the nfdv1alpha1 package as that's part of the API (or we technically *could* import it from anywhere but I think that's a bad idea because that other package then essentially becomes a part of the API...). For the "functionality", i.e. the implementation of Rule evaluation/processing, I agree that we want to move it out of the API package.